api/src/utils/sanitize-query.ts

```diff
@@ -1,18 +1,27 @@
 import { Accountability, Aggregate, Filter, Query } from '@directus/shared/types';
 import { parseFilter, parseJSON } from '@directus/shared/utils';
 import { flatten, get, isPlainObject, merge, set } from 'lodash';
+import { getEnv } from '../env';
 import logger from '../logger';
 import { Meta } from '../types';
 
 export function sanitizeQuery(rawQuery: Record<string, any>, accountability?: Accountability | null): Query {
 	const query: Query = {};
 
+	const env = getEnv();
+	const hasMaxLimit = 'MAX_QUERY_LIMIT' in env && env.MAX_QUERY_LIMIT !== -1;
 	if (rawQuery.limit !== undefined) {
 		const limit = sanitizeLimit(rawQuery.limit);
 
 		if (typeof limit === 'number') {
-			query.limit = limit;
+			if (limit === -1 && hasMaxLimit) {
+				query.limit = Number(env.MAX_QUERY_LIMIT);
```

Comment:
Does this still allow for disabling the limit altogether? I think we have to add an additional `&& env.MAX_QUERY_LIMIT !== Infinity` 🤔

Reply:
Setting infinity from a string environment variable can be tricky. The current disabled state is `MAX_QUERY_LIMIT=-1` (which is also the default as that is current behavior) in which case `hasMaxLimit` will be `false` stopping the limit from being altered by this new logic.

Reply:
> Setting infinity from a string environment variable can be tricky.

Why? We're expecting the string to represent a number, and `Number('Infinity')` correctly parses to JS `Infinity`:

<img width="161" alt="CleanShot 2023-02-10 at 13 59 43@2x" src="https://user-images.githubusercontent.com/9141017/218175455-6dd9404d-e25a-4a04-a07d-f83e2b50bb75.png">

Reply:
So you can, but why introduce a new value for that when we've always used -1 for that functionality

Reply:
The env var `MAX_QUERY_LIMIT` controls what `-1` means in the context of the limit parameter, setting `MAX_QUERY_LIMIT` to `-1` means "-1 means -1", whereas `MAX_QUERY_LIMIT=Infinity` means "-1 means Infinity". The latter is a little clearer to me 🙂